src/box/tuple.c

```diff
@@ -177,122 +210,80 @@ tuple_next(struct tuple_iterator *it)
 	return NULL;
 }
 
-/** {{{ Bigref - allow tuple reference counter to be > 2^16 */
-
-enum {
-	BIGREF_FACTOR = 2,
-	BIGREF_MAX = UINT32_MAX,
-	BIGREF_MIN_CAPACITY = 16,
-	/**
-	 * Only 15 bits are available for bigref list index in
-	 * struct tuple.
-	 */
-	BIGREF_MAX_CAPACITY = UINT16_MAX >> 1
-};
-
 /**
- * Container for big reference counters. Contains array of big
- * reference counters, size of this array and number of non-zero
- * big reference counters. When reference counter of tuple becomes
- * more than 32767, field refs of this tuple becomes index of big
- * reference counter in big reference counter array and field
- * is_bigref is set true. The moment big reference becomes equal
- * 32767 it is set to 0, refs of the tuple becomes 32767 and
- * is_bigref becomes false. Big reference counter can be equal to
- * 0 or be more than 32767.
+ * Reference count that is uploaded or acquires to/from external storage
+ * in one iteration.
  */
-static struct bigref_list {
-	/** Free-list of big reference counters. */
-	uint32_t *refs;
-	/** Capacity of the array. */
-	uint16_t capacity;
-	/** Index of first free element. */
-	uint16_t vacant_index;
-} bigref_list;
+enum { TUPLE_UPLOAD_REFS = TUPLE_REF_MAX / 2 + 1 };
 
-/** Initialize big references container. */
-static inline void
-bigref_list_create(void)
+/** Convenient helper for hash table. */
+static uint32_t
+tuple_ref_get_uploaded_refs(struct tuple *tuple)
 {
-	memset(&bigref_list, 0, sizeof(bigref_list));
+	if (!tuple->has_uploaded_refs)
+		return 0;
+	mh_int_t pos = mh_uploaded_refs_find(tuple_uploaded_refs, tuple, 0);
+	assert(pos != mh_end(tuple_uploaded_refs));
+	struct tuple_uploaded_refs *uploaded =
+		mh_uploaded_refs_node(tuple_uploaded_refs, pos);
+	assert(uploaded->tuple == tuple);
+	assert(uploaded->refs >= TUPLE_UPLOAD_REFS);
+	assert(uploaded->refs % TUPLE_UPLOAD_REFS == 0);
```

Comment:
Why not store a multiple of TUPLE_UPLOAD_REFS in uploaded->refs instead of restricting ->refs to be a multiple of TUPLE_UPLOAD_REFS?

Reply:
I don't want to make extra operations, especially division.